#19246 Auto and manual reset and `BackoffOptions`.

RayRoestenburg opened the pull request:
Also moved `TransparantExponentialBackoffSupervisor` to `akka.pattern` (renamed to `BackoffOnRestartSupervisor`)

I'm going to add more tests, but please have a look if you agree with the direction I've taken.
also I haven't contributed for a while, so if I break any rules, let me know.

RayRoestenburg committed:
#19246 Auto and manual reset and `BackoffOptions`.

Also moved `TransparantExponentialBackoffSupervisor` to `akka.pattern` (renamed to `BackoffOnRestartSupervisor`)

RayRoestenburg commented:
Oh and I also still have to update the docs.

ktoso commented:
OK TO TEST

ktoso commented:
Great, thanks – will give it a look soon :)

ktoso commented:
Refs #19246

henrymai commented on akka-actor-tests/src/test/scala/akka/pattern/BackoffSupervisorSpec.scala:
```diff
@@ -70,5 +85,39 @@ class BackoffSupervisorSpec extends AkkaSpec with ImplicitSender {
       }
     }
+
+    "support manual reset" in {
```
Maybe parameterize this whole block so that both onStop() and onFailure() can both be tested

RayRoestenburg replied:
I'll add more tests.

henrymai commented on akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala:
```diff
@@ -0,0 +1,115 @@
+        system.scheduler.scheduleOnce(minBackoff, self, ResetRestartCount(numRestarts + 1))
+      }
+      become(watching(childRef, numRestarts + 1))
+    case _ ⇒
```
Remove this case

RayRoestenburg replied:
I'll change it to 
```
case msg => unhandled(msg)
```
So it gets reported and not blackholed.

henrymai commented on akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala:
```diff
@@ -0,0 +1,115 @@
+      val delay = akka.pattern.BackoffSupervisor.calculateDelay(
+        numRestarts, minBackoff, maxBackoff, randomFactor)
+      system.scheduler.scheduleOnce(delay, self, StartChild)
+      become(waitingToStart(numRestarts, true))
```
For the case that reset is ManualReset, do we still actually want to schedule the automatic counter reset?

RayRoestenburg replied:
FYI AFK till 28th will respond then.
On Fri, 25 Dec 2015 at 02:45, henrymai <notifications@github.com> wrote:

> In akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala
> <https://github.com/akka/akka/pull/19281#discussion_r48428078>:
>
> > +      child = Some(childRef)
> > +      watch(childRef)
> > +      if (scheduleCounterReset) {
> > +        system.scheduler.scheduleOnce(
>
> minBackoff, self, ResetRestartCount(numRestarts + 1))
> > +      }
> > +      become(watching(childRef, numRestarts + 1))
> > +    case _ ⇒
> > +  }
> > +
> > +  // Steady state
> > +  def watching(childRef: ActorRef, numRestarts: Int): Receive = {
> > +    case ScheduleRestart(`childRef`) ⇒
> > +      val delay = akka.pattern.BackoffSupervisor.calculateDelay(
> > +        numRestarts, minBackoff, maxBackoff, randomFactor)
> > +      system.scheduler.scheduleOnce(delay, self, StartChild)
> > +      become(waitingToStart(numRestarts, true))
>
> For the case that reset is ManualReset, do we still actually want to
> schedule the automatic counter reset?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/akka/akka/pull/19281/files#r48428078>.
>

RayRoestenburg replied:
Ah right, I'll change to 
```scala
become(waitingToStart(numRestarts, reset != ManualReset))
```

patriknw commented on akka-actor-tests/src/test/scala/akka/pattern/BackoffOnRestartSupervisorSpec.scala:
```diff
@@ -66,7 +65,7 @@ class TransparentExponentialBackoffSupervisorSpec extends AkkaSpec {
   }
 
-  "TransparentExponentialBackoffSupervisor" must {
+  "BackoffOnRestartSupervisor" must {
```
good name

RayRoestenburg replied:
thanks

patriknw commented on akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala:
```diff
@@ -0,0 +1,115 @@
+    case BackoffSupervisor.Reset ⇒
+      reset match {
+        case ManualReset ⇒ become(watching(childRef, 0))
+        case msg         ⇒ context.system.deadLetters.forward(msg)
```
perhaps use `unhandled` instead?

RayRoestenburg replied:
Ok

patriknw commented on akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala:
```diff
@@ -0,0 +1,115 @@
+  // Steady state
+  def watching(childRef: ActorRef, numRestarts: Int): Receive = {
+    case ScheduleRestart(`childRef`) ⇒
+      val delay = akka.pattern.BackoffSupervisor.calculateDelay(
```
`akka.pattern` not needed

RayRoestenburg replied:
Ok

patriknw commented on akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala:
```diff
@@ -0,0 +1,115 @@
+  import BackoffOnRestartSupervisor._
+  import context._
+
+  private var child: Option[ActorRef] = None
```
is this field needed?

RayRoestenburg replied:
Ah, I could just use 
```scala
child(childName)
```

patriknw commented on akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala:
```diff
@@ -0,0 +1,233 @@
+   *
+   * '''***
+   * This supervisor should not be used with `Akka Persistence` child actors.
+   * `Akka Persistence` actors, currently, shutdown unconditionally on `persistFailure()`s rather
```
the word "currently" is not needed, we have no plans on changing and software is always changing

RayRoestenburg replied:
Ok

patriknw commented on akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala:
```diff
@@ -0,0 +1,233 @@
+   * This supervisor should not be used with `Akka Persistence` child actors.
+   * `Akka Persistence` actors, currently, shutdown unconditionally on `persistFailure()`s rather
+   * than throw an exception on a failure like normal actors.
+   * [[akka.pattern.BackoffSupervisor]] should be used instead for cases where the child actor
```
the link to `akka.pattern.BackoffSupervisor` is probably obsolete

RayRoestenburg replied:
Good catch

patriknw commented on akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala:
```diff
@@ -0,0 +1,233 @@
+   * The child can stop itself and send a [[akka.actor.PoisonPill]] to the parent supervisor
+   * if it wants to do an intentional stop.
+   *
+   * Exceptions in the child are handled with the given `supervisionStrategy`. A
```
perhaps some hint about `withSupervisorStrategy` instead?

RayRoestenburg replied:
Ok

patriknw commented on akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala:
```diff
@@ -0,0 +1,233 @@
+  private[akka] def props: Props
+}
+
+private case class BackoffOptionsImpl(
```
`final` for case classes

RayRoestenburg replied:
Ah missed that one, Ok.

patriknw commented on akka-actor/src/main/scala/akka/pattern/BackoffSupervisor.scala:
```diff
@@ -96,6 +105,24 @@ object BackoffSupervisor {
+  final case class RestartCount(count: Int)
+
   private case object StartChild extends DeadLetterSuppression
   private case class ResetRestartCount(current: Int) extends DeadLetterSuppression
```
final?

RayRoestenburg replied:
Ok.

patriknw commented on akka-actor/src/main/scala/akka/pattern/BackoffSupervisor.scala:
```diff
@@ -120,58 +147,42 @@ object BackoffSupervisor {
- * stop the child, but it will be started again after the back-off duration.
- */
-final class BackoffSupervisor(
+private final class BackoffSupervisor(
```
I don't think we can make it private because of backwards compatibility.
Some link in the docs of this class would be nice.

RayRoestenburg replied:
Ok.

patriknw commented on akka-actor/src/main/scala/akka/pattern/BackoffSupervisor.scala:
```diff
@@ -201,12 +212,25 @@ final class BackoffSupervisor(
+    case Reset ⇒
+      reset match {
+        case ManualReset ⇒ restartCount = 0
+        case msg         ⇒ context.system.deadLetters.forward(msg)
```
unhandled

RayRoestenburg replied:
Ok.

patriknw commented:
only minor nitpicks, great work!

henrymai commented on akka-contrib/src/main/scala/akka/contrib/pattern/TransparentExponentialBackoffSupervisor.scala:
```diff
@@ -1,203 +0,0 @@
-/**
```
Can you have a the movement of this file from:

`akka-contrib/src/main/scala/akka/contrib/pattern/TransparentExponentialBackoffSupervisor.scala` to
`akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala`

be in a separate commit that comes before this commit (in the same pull request)?
This way we can see the changes between the two versions more easily.

RayRoestenburg replied:
Sure no problem will fix tomorrow
On Sat, 26 Dec 2015 at 21:18, henrymai <notifications@github.com> wrote:

> In
> akka-contrib/src/main/scala/akka/contrib/pattern/TransparentExponentialBackoffSupervisor.scala
> <https://github.com/akka/akka/pull/19281#discussion_r48449158>:
>
> > @@ -1,203 +0,0 @@
> > -/**
>
> Can you have a the movement of this file from:
>
>
> akka-contrib/src/main/scala/akka/contrib/pattern/TransparentExponentialBackoffSupervisor.scala
> to
> akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala
>
> be in a separate commit that comes before this commit (in the same pull
> request)?
> This way we can see the changes between the two versions more easily.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/akka/akka/pull/19281/files#r48449158>.
>

ktoso replied:
Good tip indeed, thanks @henrymai !

henrymai commented on akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala:
```diff
@@ -0,0 +1,115 @@
+        case msg         ⇒ unhandled(msg)
+      }
+    case BackoffSupervisor.GetCurrentChild ⇒
+      sender() ! BackoffSupervisor.CurrentChild(child(childName))
```
Why not just: CurrentChild(Some(childRef))

- The child can always end up being shutdown after a response to GetCurrentChild.
- We should generally not be in the watching() state when the child is shutdown. See the sequence of lns 51 - 52, 82
- Which leads to another point, which is that you may want to handle the GetCurrentChild message in the waitingToStart() state and just respond with CurrentChild(None) ?

RayRoestenburg replied:
Good points.

RayRoestenburg commented:
Tests seem to fail on generated javadoc on ci, I'll have to look into that as well.

henrymai commented:
Looks good. Just some nits and still needs a test for "support manual reset" for BackoffOptions.onFailure.

RayRoestenburg commented:
@henrymai I refactored some more, added some tests, and found there were some issues, now converging both to the same style, which I think is easier to read, (and tests pass, which is important). And I also just factored out the common behavior.
There was an annoying behavior in the first message getting black-holed on the BackoffOnRestartSupervisor because of `self ! StartChild` in `preStart`.

RayRoestenburg commented:
I did some extra cleanup and refactoring, @patriknw I'm assuming adding a private trait does not break binary compatibility?

RayRoestenburg commented:
I've also changed the docs so genjavadoc doesn't complain.

RayRoestenburg commented:
I think this should be ready now. I did refactor the code for some internal reuse gains.

ktoso commented on akka-actor/src/main/scala/akka/pattern/BackoffSupervisor.scala:
```diff
@@ -96,8 +104,26 @@ object BackoffSupervisor {
 
-  private case object StartChild extends DeadLetterSuppression
-  private case class ResetRestartCount(current: Int) extends DeadLetterSuppression
+  final case object Reset
```
Needs the same docs as the java API below.

```
Send this message to the [[BackoffSupervisor]] and it will reset the back-off.
This should be used in conjunction with `withManualReset` in [[BackoffOptions]].
```

RayRoestenburg replied:
Ok.

ktoso commented on akka-actor/src/main/scala/akka/pattern/BackoffSupervisor.scala:
```diff
@@ -96,8 +104,26 @@ object BackoffSupervisor {
+   */
+  def reset = Reset
+
+  final case object GetRestartCount
```
Needs the same docs as the Java API

```
Send this message to the [[BackoffSupervisor]] and it will reply with
[[BackoffSupervisor.RestartCount]] containing the current restart count.
```

RayRoestenburg replied:
Ok.

ktoso commented:
Failure seems like missing imports 

```
[error] /localhome/jenkinsakka/workspace/pr-validator-per-commit-jenkins/akka-docs/rst/scala/code/docs/pattern/BackoffSupervisorDocSpec.scala:91: not found: value OneForOneStrategy
[error]           OneForOneStrategy() {
[error]           ^
[error] /localhome/jenkinsakka/workspace/pr-validator-per-commit-jenkins/akka-docs/rst/scala/code/docs/pattern/BackoffSupervisorDocSpec.scala:92: not found: value SupervisorStrategy
[error]             case _: MyException ⇒ SupervisorStrategy.Restart
[error]                                   ^
[error] /localhome/jenkinsakka/workspace/pr-validator-per-commit-jenkins/akka-docs/rst/scala/code/docs/pattern/BackoffSupervisorDocSpec.scala:93: not found: value SupervisorStrategy
[error]             case _              ⇒ SupervisorStrategy.Escalate
[error]                                   ^
```

ktoso commented on akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala:
```diff
@@ -0,0 +1,234 @@
+ * var options = BackoffOptions.onFailure(childProps, childName, minBackoff, maxBackoff, randomFactor)
+ * options = options.withManualReset
+ * context.actorOf(BackoffSupervisor.props(options), name)
+ * }}}
```
let's not show the `var` style :) People can figure it out if they need to, it's not something we want to encourage though I think

RayRoestenburg replied:
Ok.

ktoso commented on akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala:
```diff
@@ -0,0 +1,234 @@
+trait BackoffOptions {
+  /**
+   * Returns a new BackoffOptions with automatic back-off reset.
+   * The back-off algorithm is reset if the child does not crash witin the specified `resetBackoff`.
```
s/witin/within/

RayRoestenburg replied:
Ok.

ktoso commented on akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala:
```diff
@@ -0,0 +1,234 @@
+
+  /**
+   * Returns a new BackoffOptions with manual back-off reset. The back-off is only reset
+   * if the child sends a `BackoffSupervisor.Reset` to its parent (the backoff-supervisor actor).
```
[[BackoffSupervisor.Reset]] (so it's navigable in scaladoc)

RayRoestenburg replied:
`[[` gave me trouble with genjavadoc, does a single `[` work for scaladoc?

ktoso replied:
hah, just noticed that `[[]] ()` above was rendered as markdown by github in above comment ;-)

Hm, I see `[[]]` may be troublesome because inner object; Ignore this comment then, not a big issue anyway

RayRoestenburg replied:
alright.

ktoso commented on akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala:
```diff
@@ -0,0 +1,234 @@
+    backoffReset match {
+      case AutoReset(resetBackoff) ⇒
+        require(minBackoff <= resetBackoff && resetBackoff <= maxBackoff)
+      case _ ⇒
```
ok to ignore here, but let's add `// ignore` so that it's obvious that it's intentional

RayRoestenburg replied:
Ok.

ktoso commented on akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala:
```diff
@@ -0,0 +1,234 @@
+
+private sealed trait BackoffType
+private final case object StopImpliesFailure extends BackoffType
+private final case object RestartImpliesFailure extends BackoffType
```
like those names :+1:

ktoso commented on akka-actor/src/main/scala/akka/pattern/BackoffSupervisor.scala:
```diff
@@ -78,6 +80,12 @@ object BackoffSupervisor {
+   * Props for creating a [[BackoffSupervisor]] actor from [[BackoffOptions]].
+   * @param options the [[BackoffOptions]] that specify how to construct a backoff-supervisor.
+   */
+  def props(options: BackoffOptions): Props = options.props
```
:+1:

RayRoestenburg replied:
Thanks

ktoso commented on akka-actor/src/main/scala/akka/pattern/BackoffSupervisor.scala:
```diff
@@ -121,59 +147,45 @@ object BackoffSupervisor {
 
-  private var child: Option[ActorRef] = None
-  private var restartCount = 0
+  // to keep binary compatibility with 2.4.1
```
good, thanks

ktoso commented on akka-actor/src/main/scala/akka/pattern/BackoffSupervisor.scala:
```diff
@@ -184,29 +196,57 @@ final class BackoffSupervisor(
+      reset match {
+        case AutoReset(resetBackoff) ⇒
+          val _ = context.system.scheduler.scheduleOnce(resetBackoff, self, ResetRestartCount(restartCount))
+        case _ ⇒
```
add an `// ignore` so we know at a glance that it's not accidental omission

RayRoestenburg replied:
Ok.

RayRoestenburg commented:
Hm, crap. what is the task to run locally to check this BTW?

On Tue, Dec 29, 2015 at 2:56 PM Konrad Malawski <notifications@github.com>
wrote:

> Failure seems like missing imports
>
> [error] /localhome/jenkinsakka/workspace/pr-validator-per-commit-jenkins/akka-docs/rst/scala/code/docs/pattern/BackoffSupervisorDocSpec.scala:91: not found: value OneForOneStrategy
> [error]           OneForOneStrategy() {
> [error]           ^
> [error] /localhome/jenkinsakka/workspace/pr-validator-per-commit-jenkins/akka-docs/rst/scala/code/docs/pattern/BackoffSupervisorDocSpec.scala:92: not found: value SupervisorStrategy
> [error]             case _: MyException ⇒ SupervisorStrategy.Restart
> [error]                                   ^
> [error] /localhome/jenkinsakka/workspace/pr-validator-per-commit-jenkins/akka-docs/rst/scala/code/docs/pattern/BackoffSupervisorDocSpec.scala:93: not found: value SupervisorStrategy
> [error]             case _              ⇒ SupervisorStrategy.Escalate
> [error]                                   ^
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/akka/akka/pull/19281#issuecomment-167783195>.
>

ktoso commented on akka-actor/src/main/scala/akka/pattern/BackoffSupervisor.scala:
```diff
@@ -184,29 +196,57 @@ final class BackoffSupervisor(
+    case Reset ⇒
+      reset match {
+        case ManualReset ⇒ restartCount = 0
+        case msg         ⇒ unhandled(msg)
```
good (was thinking if logging warnings would be better, but unhandled is OK)

ktoso commented on akka-docs/rst/general/supervision.rst:
```diff
@@ -203,25 +203,26 @@ terminates a child by way of the ``system.stop(child)`` method or sending a
 
-The following Scala snippet shows how to create a backoff supervisor which will start the given echo actor
-in increasing intervals of 3, 6, 12, 24 and finally 30 seconds:
+.. [#] A failure can be indicated in two different ways; by an actor terminating or crashing.
```
slippery wording here - technically when it crashes and stops "it terminates". Instead of terminating say "stops" here?

RayRoestenburg replied:
Ok.

ktoso commented:
> Hm, crap. what is the task to run locally to check this BTW?

Either of:

```
sbt shell
> test
# or
> project akka-docs
> test
```

RayRoestenburg commented:
@ktoso Aha thanks.

ktoso commented on akka-actor-tests/src/test/scala/akka/pattern/BackoffSupervisorSpec.scala:
```diff
@@ -47,27 +64,92 @@ class BackoffSupervisorSpec extends AkkaSpec with ImplicitSender {
+        val restartingStrategy = OneForOneStrategy() {
+          case _: TestException ⇒ SupervisorStrategy.Restart
+        }
+        assertManualReset(create(onStopOptions(ManualChild.props(testActor)).withManualReset.withSupervisorStrategy(stoppingStrategy)))
```
break into multiple lines

RayRoestenburg replied:
Ok.

ktoso commented on akka-actor-tests/src/test/scala/akka/pattern/BackoffSupervisorSpec.scala:
```diff
@@ -47,27 +64,92 @@ class BackoffSupervisorSpec extends AkkaSpec with ImplicitSender {
+        awaitAssert {
+          supervisor ! BackoffSupervisor.GetCurrentChild
+          // new instance
+          expectMsgType[BackoffSupervisor.CurrentChild].ref.get should !==(c1)
```
maybe also check that the restarts count was incremented by sending `GetRestartCount`?

RayRoestenburg replied:
Ok.

ktoso commented:
Looks awesome, thanks for the contrib @RayRoestenburg!
I think I ran out of nitpicks, and the core looks good.

Let's get it to compile and pass tests then :)

RayRoestenburg commented:
@ktoso Ok will do. Hm, I forgot Java does not like companion objects, so I renamed the `BackoffOptions` object to `Backoff`, so you get `Backoff.onStop` and `Backoff.onFailure` (which is not bad).

ktoso commented:
It should work just fine with a companion as long as it's a real one – as in `class + object`, `trait + object` indeed causes trouble AFAIR

RayRoestenburg commented:
Yeah that's what I meant, it was `trait + object`, so not a real one..

ktoso commented:
Ok, I'll give the current names a look then :)

ktoso commented:
`Backoff.on...` reads nice in any case :) LGTM from my side.

RayRoestenburg commented:
Yeah accidentally a better name :smile:. I'm really bad at naming most of the time. Cool. now lets see if it finally passes all the docs etc..

RayRoestenburg commented:
@ktoso ehh, help?

> akka-actor: Binary compatibility check failed!

Where can I get more information?

ktoso commented:
You can run `mimaReportBinaryIssues` to see the problem. If it's just internals we can add a filter, if it's actual API we need to work around it such that it's not breaking binary compat.

RayRoestenburg commented:
@ktoso thanks I'll have a look.

ktoso commented:
Failures are:

```
[error]  * synthetic method akka$pattern$BackoffSupervisor$$child_=(scala.Option)Unit in class akka.pattern.BackoffSupervisor does not have a correspondent in new version
[error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.pattern.BackoffSupervisor.akka$pattern$BackoffSupervisor$$child_=")
[error]  * synthetic method akka$pattern$BackoffSupervisor$$restartCount()Int in class akka.pattern.BackoffSupervisor does not have a correspondent in new version
[error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.pattern.BackoffSupervisor.akka$pattern$BackoffSupervisor$$restartCount")
[error]  * synthetic method akka$pattern$BackoffSupervisor$$restartCount_=(Int)Unit in class akka.pattern.BackoffSupervisor does not have a correspondent in new version
[error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.pattern.BackoffSupervisor.akka$pattern$BackoffSupervisor$$restartCount_=")
[error]  * synthetic method akka$pattern$BackoffSupervisor$$child()scala.Option in class akka.pattern.BackoffSupervisor does not have a correspondent in new version
[error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.pattern.BackoffSupervisor.akka$pattern$BackoffSupervisor$$child")
[error]  * class akka.pattern.BackoffSupervisor#ResetRestartCount is declared final in new version
[error]    filter with: ProblemFilters.exclude[FinalClassProblem]("akka.pattern.BackoffSupervisor$ResetRestartCount")
```

I'm looking into it (which ones we can ignore and which not)

RayRoestenburg commented:
@ktoso I'm also looking into it, I can declare the `ResetRestartCount` as not final (easy), the others are a result of refactoring into `HandleBackoff`, the `child` and `restartCount` were public vars in the previous version (since it's an `Actor` no one can touch them so that makes sense why it was that way) so maybe those can be filtered out? Otherwise I'll have to leave them in as dummies and change the names internally, which is not that nice.

RayRoestenburg commented:
Hm, doesn't seem to be as easy as that, even if I add those vars back in. @ktoso let me now if you can filter this out.

ktoso commented:
The class needs to be un-final-ized yeah. 
The rest, in fact, we can ignore all these since they are about BackoffSupervisor which is an `Actor` and is `final` anyway.

Please add the filters as the above error message lists them to `project/MiMa.scala`, like so:

```
ProblemFilters.exclude[MissingMethodProblem]("akka.remote.InvalidAssociation.this"),

        // #19281 BackoffSupervisor updates
        ProblemFilters.exclude[FinalClassProblem]("akka.pattern.BackoffSupervisor$ResetRestartCount")
          ...
```

Thanks!

RayRoestenburg commented:
@ktoso done. I've made the `ResetRestartCount` non-final I think that is what you meant right? All others are ignored using the `ProblemFilters`.

ktoso commented:
Yup, thanks!
On Dec 29, 2015 9:38 PM, "akka-ci" <notifications@github.com> wrote:

> Test PASSed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/akka/akka/pull/19281#issuecomment-167873703>.
>

ktoso commented:
Patrik reviewed and liked it and LGTM from my side as well so merging.
Really excellent improvement, thanks Ray!

RayRoestenburg commented:
@ktoso you're welcome!
